zerver/tests/test_events.py

```diff
@@ -536,6 +537,31 @@ def test_huddle_send_message_events(self) -> None:
             lambda: self.send_huddle_message(self.example_user("cordelia"), huddle, "hola"),
         )
 
+    def test_user_creation_events_on_sending_messages(self) -> None:
+        self.set_up_db_for_testing_user_access()
+        polonius = self.example_user("polonius")
+        cordelia = self.example_user("cordelia")
+
+        self.user_profile = polonius
+        events = self.verify_action(
+            lambda: self.send_personal_message(cordelia, polonius, "hola"), num_events=2
+        )
+        check_direct_message("events[0]", events[0])
+        check_realm_user_add("events[1]", events[1])
+        self.assertEqual(events[1]["person"]["user_id"], cordelia.id)
+
+        othello = self.example_user("othello")
+        desdemona = self.example_user("desdemona")
+
+        events = self.verify_action(
+            lambda: self.send_huddle_message(othello, [polonius, desdemona], "hola"), num_events=3
+        )
+        check_direct_message("events[0]", events[0])
+        check_realm_user_add("events[1]", events[1])
+        check_realm_user_add("events[2]", events[2])
```

Comment:
The order of events here in tests is not same as in we would get in production environment because `notify_created_user` uses `send_event_on_commit` and the tests code execute the `on_commit` calls at last.

Reply:
I think actually the same thing will happen in production -- the event sent by `send_event_on_commit` will happen last. We likely need to change the main direct message event to be sent via `send_event_on_commit` rather than `send_event` to fix this bug.